Event Queues (#2608)
cap-js/docs: guides/events/event-queues.md

- `<event>/#failed` — fires when the message becomes a dead letter (after all retries are exhausted).

**Example:** After *xflights* successfully processes a `BookingCreated` event, the *xtravels* application replicates the booking confirmation back into its own database. If the booking fails, the application updates the local `Bookings` row to surface the error in its UI.

```js
const xflights = await cds.connect.to('xflights')

// Called when the queued booking succeeds
xflights.after('BookingCreated/#succeeded', async (result, req) => {
  console.log('Flight booked successfully:', result)
  // Replicate booking details from remote
})

// Called when the queued booking fails after max retries
xflights.after('BookingCreated/#failed', async (error, req) => {
  console.log('Flight booking failed:', error)
  // Trigger compensation logic
})
```

This is also the foundation for [SAGA-style](https://microservices.io/patterns/data/saga.html) compensation across distributed systems: two-phase commits aren't possible once an outboxed call has gone out, so you keep consistency by reacting to outcomes and compensating where needed.

> [!note] Node.js only

> > Callback events `#succeeded` and `#failed` are currently available in Node.js only. Java doesn't have an equivalent yet, but it's on the roadmap.

Reviewer: This feature is requested by stakeholders frequently and would solve many workarounds.

The outbox fixes this. Wrap the remote service in `cds.queued()` (Node.js) or `OutboxService.outboxed()` (Java) and dispatch as before. The call is now persisted within the current transaction and sent after commit:

::: code-group
```js [Node.js]
const xflights = await cds.connect.to('xflights')
const qd_xflights = cds.queued(xflights)

this.after('CREATE', 'Bookings', async (_, req) => {
  const { flight_ID: flight, flight_date: date } = req.data
  // Persisted within the current transaction, sent after commit      // [!code ++]
  await qd_xflights.send('POST', 'BookingCreated', { flight, date })  // [!code ++]
})
```
```java [Java]
@Autowired @Qualifier("XFlightsOutbox")
OutboxService outbox;

@Autowired @Qualifier(CqnService.DEFAULT_NAME)
CqnService xflights;

@After(event = CqnService.EVENT_CREATE, entity = Bookings_.CDS_NAME)
void notifyXFlights(List<Bookings> bookings) {
  AsyncCqnService outboxedXFlights = AsyncCqnService.of(xflights, outbox);
  bookings.forEach(b -> outboxedXFlights.emit("BookingCreated",

>     Map.of("flight", b.getFlightId(), "date", b.getFlightDate())));

Reviewer 2: ```suggestion
@Autowired
OutboxService outbox;

@Autowired
TravelService xflights;

@After(event = CqnService.EVENT_CREATE, entity = Bookings_.CDS_NAME)
void notifyXFlights(List<Bookings> bookings) {
  xflights = outbox.outboxed(xflights);
  bookings.forEach(b -> xflights.bookingCreated(...));
```

- A **direct** call returns the remote service's result (or error) before the local transaction commits.
- A **queued** call writes the message to the queue inside the local transaction and returns. The actual remote dispatch happens after commit, in the background.

> [!warning] Queued calls discard the direct return value
> A queued service persists the request and returns after the message is stored, not after the remote operation finishes. Any return value from `send()` or `run()` is therefore not available to the caller. To act on the outcome, register a [callback handler](#callbacks) on `#succeeded` or `#failed`.

> [!tip] `await` is still needed
> Even though processing is asynchronous, you still need to `await` because the message is written to the database within the current transaction.

In Java, you can also wrap a service at runtime through the service catalog rather than wiring through Spring:

```java
OutboxService outbox = runtime.getServiceCatalog()
    .getService(OutboxService.class, "XFlightsOutbox");
CqnService xflights = runtime.getServiceCatalog()
    .getService(CqnService.class, "xflights");

AsyncCqnService queued = AsyncCqnService.of(xflights, outbox);

> queued.emit("BookingCreated", Map.of("flight", "AA017", "date", "2026-07-15"));

Reviewer 2: ```suggestion
OutboxService outbox = runtime.getServiceCatalog()
    .getService(OutboxService.class, "XFlightsOutbox");
TravelService xflights = runtime.getServiceCatalog()
    .getService(TravelService.class, "xflights");

xflights = outbox.outboxed(xflights);
xflights.bookingCreated(...);
```

```

To get the original synchronous service from a queued proxy:

::: code-group
```js [Node.js]
const xflights = cds.unqueued(qd_xflights)
```
```java [Java]
CqnService xflights = OutboxService.unboxed(outboxedXFlights);
```
:::

### By Configuration

To outbox a service centrally, without touching handler code, set a flag on its configuration. Every call from your handlers is then queued automatically.

::: code-group
```json [Node.js — package.json]
{
  "cds": {
    "requires": {
      "messaging": {
        "outboxed": true
      }
    }
  }
}
```
```yaml [Java — application.yaml]
cds:
  outbox:
    services:
      DefaultOutboxUnordered:

>         maxAttempts: 10

Reviewer 2: In Java, we do not support enabling outboxing via configuration for arbitrary services. Only Messaging and Audit Log services support enabling or disabling outboxing through configuration.

`cds.messaging.services.<service>.outbox.enabled`
`cds.auditLog.outbox.enabled`

Author: → https://github.com/capire/docs/pull/2608/changes/ecdf84db2eae320d9073df957194e7deb42d3dea

This brings two advantages:

- **Quick acknowledgment**: the broker no longer waits for your processing to complete, which keeps consumer throughput high under load.
- **Controlled processing rate**: if a burst of messages arrives, they are queued in your database and processed at a controlled pace.

> [!note] Especially useful when broker redelivery doesn't fit
> Some message brokers don't allow redelivery or payload correction. Others have fixed redelivery timeouts that expire when your processing legitimately takes longer than the broker's window. With the inbox, the broker's job ends at acknowledgement and failures are handled inside your app via the [dead letter queue](#dead-letter-queue), where you have full control over retry timing, payload correction, and discard.

Enable the inbox in your configuration:

::: code-group
```json [Node.js — package.json]
{
  "cds": {
    "requires": {
      "messaging": {
        "inboxed": true
      }
    }
  }
}
```
```yaml [Java — application.yaml]
cds:
  messaging:
    services:
      - name: messaging-name
        inbox:

>           enabled: true

Reviewer 2: The inbox is disabled by default and must be explicitly enabled, as shown in the example. It may also be worth explaining that the inbox uses the default ordered outbox by default. This clarifies how inbound messages are processed and ensures that message handling semantics are understood.

In addition, users can override the default outbox by specifying a custom outbox via the cds.messaging.services.<key>.inbox.name configuration property.

Author: please add to java guide (maybe as part of "Default Outbox Services"?), preferably via follow-up.

> With inboxing enabled, the broker considers the message delivered as soon as your app stores it.
> If later processing fails, recovery no longer happens in the broker; it happens in your application's retry and dead letter queue flow.

## Scheduled Tasks

Event queues are not limited to outbound calls and messaging.
You can schedule arbitrary work such as data replication, cache refresh, or garbage collection.

A scheduled task is identified by its event name and exists only once: a subsequent `schedule()` call with the same name overwrites the previous schedule (tasks are upserted, not deduplicated). This makes scheduling idempotent, which is convenient during application startup, where the same registration code runs on every boot.

**Example:** Replicate airport master data from the *xflights* service every 10 minutes.

::: code-group
```js [Node.js]
const xflights = await cds.connect.to('xflights')
await xflights.schedule('replicate', { entity: 'Airports' }).every('10m')
```
```java [Java]
@Autowired
OutboxService outbox;

@Autowired
RemoteService xflights;

Schedulable.of(xflights, outbox)
  .scheduled(Schedule.create().every(Duration.ofMinutes(10)))

>   .emit("replicate", Map.of("entity", "Airports"));

Reviewer 2: ```suggestion
@Autowired
OutboxService outbox;

@Autowired
TravelService xflights;

Schedulable.of(xflights, outbox)
  .scheduled(Schedule.create().every(Duration.ofMinutes(10)))
  .replicateTravels(...);
```

module.exports = class OutboxDeadLetterQueueService extends cds.ApplicationService {
  async init() {
    this.before('READ', 'DeadOutboxMessages', function (req) {
      const { maxAttempts } = cds.env.requires.queue
      req.query.where('attempts >= ', maxAttempts)
    })
    await super.init()
  }
}
```
```java [Java — DeadOutboxMessagesHandler.java]
@Component
@ServiceName(OutboxDeadLetterQueueService_.CDS_NAME)
public class DeadOutboxMessagesHandler implements EventHandler {

  private final PersistenceService db;

  public DeadOutboxMessagesHandler(
      @Qualifier(PersistenceService.DEFAULT_NAME) PersistenceService db) {
    this.db = db;
  }

  @Before(event = CqnService.EVENT_READ, entity = DeadOutboxMessages_.CDS_NAME)
  public void addDeadEntryFilter(CdsReadEventContext context) {
    Optional<Predicate> outboxFilters = createOutboxFilters(context.getCdsRuntime());
    outboxFilters.ifPresent(filter -> {
      CqnSelect modified = copy(context.getCqn(), new Modifier() {
        @Override
        public CqnPredicate where(Predicate where) {
          return filter.and(where);
        }
      });
      context.setCqn(modified);
    });
  }

> }

Reviewer 2: In contrast to our previous explanation in the Java section, we also provide the addDeadEntryFilter method here.

```java
private Optional<Predicate> createOutboxFilters(CdsRuntime runtime) {
        CdsProperties.Outbox outboxConfigs = runtime.getEnvironment().getCdsProperties().getOutbox();

        return runtime.getServiceCatalog().getServices(OutboxService.class)
                 .map(service -> {
                     OutboxServiceConfig config = outboxConfigs.getService(service.getName());
                     return CQL.get(Messages.TARGET).eq(service.getName())
                              .and(CQL.get(Messages.ATTEMPTS).ge(config.getMaxAttempts()));
                 })
                 .reduce(Predicate::or);
    }
```

Author: i need more context for this one

Author: → https://github.com/capire/docs/pull/2608/commits/7afc9f408296709fb9eb459b3d5a29690448ff4f

- **Outbox**: defer outbound calls to remote services and emit messages to message brokers until the transaction succeeds.
- **Inbox**: acknowledge inbound messages immediately and process them asynchronously.
- **Scheduled Tasks**: run periodic or delayed work such as data replication.

### Pub/Sub vs. Event Queues

These are sometimes confused but solve different problems.

**Pub/sub**, typically realized through a message broker, addresses *loosely coupled microservices*. A producer publishes events without knowing who consumes them; consumers subscribe by topic. The unit of trust is the broker.

**Event queues** address *asynchronous workload processing within one service*. They turn a piece of work into a database row that survives commit, restart, and retry, then dispatch it later: to the same service in process, to a remote service, or to a message broker. The unit of trust is the database transaction.

The two patterns complement each other: when the dispatch target *is* a message broker, the event queue is the transactional bridge that makes pub/sub safe across the local commit. The [Inbox](#inbox) does the mirror image on the receiving side.

> [!note] Related patterns

> > [*Event Sourcing*](https://microservices.io/patterns/data/event-sourcing.html) solves the same atomic-state-change-and-publish problem by making an append-only event log the source of truth. Event queues persist messages only until processed and then delete them — they're a transactional bridge to remote systems, not the system of record.

Reviewer 3: I stumbled here and gave it a try to rephrase
```suggestion
> [*Event Sourcing*](https://microservices.io/patterns/data/event-sourcing.html) solves the same atomic-state-change-and-publish problem by establishing a source of truth through an append-only event log. Event queues persist messages only until processed and then delete them — they're a transactional bridge to remote systems, not the system of record.
```

- A **queued** call writes the message to the queue inside the local transaction and returns. The actual remote dispatch happens after commit, in the background.

> [!warning] Queued calls discard the direct return value
> A queued service persists the request and returns after the message is stored, not after the remote operation finishes. Any return value from `send()` or `run()` is therefore not available to the caller. To act on the outcome, register a [callback handler](#callbacks) on `#succeeded` or `#failed`.

> [!tip] `await` is still needed
> Even though processing is asynchronous, you still need to `await` because the message is written to the database within the current transaction.

In Java, you can also wrap a service at runtime through the service catalog rather than wiring through Spring:

```java
OutboxService outbox = runtime.getServiceCatalog()
    .getService(OutboxService.class, "XFlightsOutbox");
TravelService xflights = runtime.getServiceCatalog()
    .getService(TravelService.class, "xflights");

xflights = outbox.outboxed(xflights);
xflights.bookingCreated(...);

> ```

Reviewer 3: Are these two things the only runtime-specific things in this section? Everything else is valid for Java and Node.js?

Reviewer 3: We could also add the snippet to the code group above with the "header": Java w/o Spring
WDYT?

Author: which two things specifically?
thinking about it, maybe we should move the "wrap a service at runtime" stuff to the java guide?  it doesn't really add value in the generic guide, right?

Reviewer 3: the `await`and the "wrap a service" where the things I meant

Author: ah, ok. my suggestion would be:
- "> [!tip] Node.js: `await` is still needed"
- move "wrap a service" to the java guide

Reviewer 3: @dimamost WDYT about moving the "wrap a service" into the Java guide and have a learn more link here?

Author: @renejeglinsky → d7b590f6ae82d714fa3872f087d332c7f92b4aa7
added to code block as "Java w/o @Autowired" after all as there was no immediate fit in the java guide

Reviewer 2: > @dimamost WDYT about moving the "wrap a service" into the Java guide and have a learn more link here?

I don't mind.

::: code-group
```js [Node.js]
const xflights = cds.unqueued(qd_xflights)
```
```java [Java]
CqnService xflights = OutboxService.unboxed(outboxedXFlights);
```
:::

### By Configuration

To outbox a required service centrally, without touching handler code, set a flag on its configuration. Every call from your handlers is then queued automatically.

::: code-group
```json [Node.js — package.json]
{
  "cds": {
    "requires": {
      "messaging": {
        "outboxed": true
      }
    }
  }
}
```
:::

This is the typical setup for **technical services**, such as messaging and audit logging, where every emit must be durable. CAP enables it by default for those services (see [*Auto-Outboxed Services*](#auto-outboxed-services) below).

For **business services**, however, a class-level flag is usually too coarse. Remote integrations called from domain handlers typically need *some* calls outboxed, for example, the post-commit notification to *xflights*, while others stay synchronous (a read-through query, a probe before commit). For finer control, prefer the programmatic path with `cds.queued()` or `srv.schedule()`.

> [!note] Node.js only

> > Outboxing required services by configuration is available in Node.js only.

Reviewer 3: Multitenancy: How about using outboxing in a Node.js sidecar with a Java Project? Does that make sense?

Author: in a node.js sidecar, node.js rules apply :) but that's always the case. it's a node.js cap app.

Reviewer 3: So there's no conflict in outboxing by config in the sidecar and programmatic outboxing in the main project?

Author: no. btw, they also would not be outboxing the same things (but also no issue if they would). in the sidecar, you'd maybe outbox some things during tenant onboarding or so, but nothing business logic related.

This pattern is widely known as the *'Transactional Outbox'*, but CAP's event queues go beyond outbound messages. They cover four use cases:

- **Outbox** — defer outbound calls to remote services until the transaction succeeds.
- **Inbox** — acknowledge inbound messages immediately and process them asynchronously.
- **Background Tasks** — schedule periodic or delayed tasks such as data replication.
- **Callbacks** — react to completed or failed tasks, enabling SAGA-like compensation patterns.

## Quick Start

Event queues are enabled by default — there's nothing to install or activate. Use a queued service when a side effect must only happen after the current transaction commits.

```js
const xflights = await cds.connect.to('xflights')
const qd_xflights = cds.queued(xflights)

this.after('CREATE', 'Travels', async (_, req) => {
  await qd_xflights.send('bookFlight', { travelId: req.data.ID })
})
```

This stores the flight booking request in the database together with the travel creation. CAP dispatches it later in the background. If the transaction rolls back, no booking request is sent.

> The `xflights` connection in the example stands in for any remote service you've configured under `cds.requires`. The complete setup, including the *XTravels* application and the *xflights* service it consumes, lives in the [@capire/xtravels](https://github.com/capire/xtravels) sample.

Reviewer 4: *XTravels* -> *xtravels*

CqnService xflights;

@After(event = CqnService.EVENT_CREATE, entity = Travels_.CDS_NAME)
void notifyXFlights(List<Travels> travels) {
  AsyncCqnService outboxedXFlights = AsyncCqnService.of(xflights, outbox);
  travels.forEach(t -> outboxedXFlights.emit("bookFlight", Map.of("travelId", t.getId())));
}
```
:::

```js
// Anti-pattern: remote side effect happens before local commit is safe
this.after('CREATE', 'Travels', async (_, req) => {
  await xflights.send('bookFlight', { travelId: req.data.ID })
})
```

If the surrounding transaction later fails, the external booking may already exist although the local travel record was rolled back.

[See the *XTravels* sample for a comparable scenario.](https://github.com/capire/xtravels){.learn-more}

Some services are outboxed automatically, including `cds.MessagingService` and `cds.AuditLogService`.
You don't need to call `cds.queued()` or configure anything extra for these — they use the persistent queue by default.

[Learn more about auto-outboxed services in Node.js.](../../node.js/event-queues#queueing-a-service){.learn-more}
[Learn more about the outbox in Java.](../../java/event-queues){.learn-more}

### Inbox

> The inbox mirrors the *'Outbox'* pattern for inbound messages.

Reviewer 4: Above (line 115): This pattern is widely known as the *'Transactional Outbox'*,

When a message arrives from a broker, the messaging service immediately persists it to the database, acknowledges it to the broker, and schedules its processing.

This brings two advantages:

- **Quick acknowledgment** — the broker doesn't have to wait for your processing to complete, which keeps consumer throughput high under load.
- **Flatten the curve** — if a burst of messages arrives, they are queued in your database and processed at a controlled pace.

> [!note] Especially useful when brokers don't support redelivery
> Some message brokers do not allow retriggering delivery or correcting message payloads.
> With the inbox, failures are handled inside your app via the [dead letter queue](#dead-letter-queue), where you have full control over retry and correction.

Enable the inbox in your configuration:

::: code-group

> ```json [Node.js — package.json]

Reviewer 4: Consider adding a brief note highlighting differences between inboxed and inbox.enabled

{
  "cds": {
    "requires": {
      "messaging": {
        "inboxed": true
      }
    }
  }
}
```
```yaml [Java — application.yaml]
cds:
  messaging:
    services:
      - name: messaging-name
        inbox:
          enabled: true
```
:::

::: warning Inboxing shifts failure handling to your application
With inboxing enabled, the broker considers the message delivered as soon as your app stores it.
If later processing fails, recovery no longer happens in the broker; it happens in your application's retry and dead letter queue flow.
:::

### Background Tasks

Event queues are not limited to messaging.
You can schedule arbitrary background tasks such as data replication, cache refresh, or garbage collection.

> **Example:** Replicate airport master data from the xflights service every 10 minutes.

Reviewer 4: italic *xflights*

await xflights.unschedule.task('replicate')
```

The event name doubles as the task name. A subsequent call with the same name overwrites the previous schedule (tasks are upserted, not deduplicated). This is convenient for idempotent registration during application startup.

::: tip Real-world example: data federation
The [data federation guide](../integration/data-federation) uses `srv.schedule().every()` to implement polling-based replication, fetching incremental updates from remote services on a regular interval.
:::

### Callbacks (SAGA Patterns) <Beta />

In distributed transactions, you often need to react when an asynchronous step completes or fails.
Event queues support this with `#succeeded` and `#failed` callback events, enabling compensation logic similar to SAGA patterns.

**Example:** After successfully creating a flight booking through *xflights*, the *XTravels* application replicates the full booking back into its own database. If the booking fails, the application updates the local `Travels` row to surface the error in its UI.

::: code-group
```js [Node.js]
const xflights = await cds.connect.to('xflights')

// Called when the queued booking succeeds

> xflights.after('bookFlight/#succeeded', async (result, req) => {

Reviewer 4: it's not explained what result contains: 
The return value of the event handler OR
The response object

AsyncCqnService queued = AsyncCqnService.of(xflights, outbox);
queued.emit("bookFlight", Map.of("travelId", "T-42"));
```
:::

A queued call changes _when_ work happens and _what the caller can expect back_:

- A **direct** call returns the remote service's result (or error) and only then commits the local transaction.
- A **queued** call writes the message to the queue inside the local transaction and returns. The actual remote dispatch happens after commit, in the background.

> [!warning] Queued calls discard the direct return value
> A queued service persists the request and returns after the message is stored, not after the remote operation finishes.
> Any return value from `send()` or `run()` is therefore not available to the caller. To act on the outcome, register a [callback handler](#callbacks-saga-patterns) on `#succeeded` or `#failed`.

::: tip `await` is still needed
Even though processing is asynchronous, you still need to `await` because the message is written to the database within the current transaction.
:::

To unwrap a queued service back to its synchronous original:

::: code-group
```js [Node.js]
const xflights = cds.unqueued(qd_xflights)
```
```java [Java]

> CqnService xflights = outbox.unboxed(outboxedXFlights);

Reviewer 4: The unboxed() method appears to be a static method on OutboxService:

CqnService xflights = OutboxService.unboxed(outboxedXFlights);

cds:
  outbox:
    services:
      XFlightsOutbox:
        maxAttempts: 10
```
:::

#### Configuring the Queue

The persistent queue is enabled by default. Messages are stored in a database table within the current transaction.

> [!note] Defaults differ between stacks
> Node.js enables the **persistent** queue for every queued service by default. Java enables the persistent outbox for `cds.MessagingService` and `cds.AuditLogService` only; other services use the in-memory outbox unless you opt them in via `cds.requires.outbox.kind: persistent-outbox`. The Java configuration sample below already does that.

::: code-group
```json [Node.js — package.json]
{
  "cds": {
    "requires": {
      "scheduling": {},
      "queue": {
        "maxAttempts": 20,
        "chunkSize": 10
      }
    }
  }
}
```
```yaml [Java — application.yaml]
cds:
  outbox:
    services:
      DefaultOutboxOrdered:
        maxAttempts: 10
        ordered: true
      DefaultOutboxUnordered:
        maxAttempts: 10
        ordered: false
```
:::

::: details Queue and scheduling options for Node.js

`cds.requires.queue`:

| Option | Default | Description |
|--------|---------|-------------|

> | `maxAttempts` | `20` | Maximum retries before a message becomes a dead letter |

Reviewer 4: maxAttempts is different for node and java, add to "Stack Differences at a Glance" table

| `chunkSize` | `10` | Number of messages to process per batch |
| `storeLastError` | `true` | Store error information of the last failed attempt |

> | `timeout` | `"1h"` | Time after which a `processing` message is considered abandoned |

Reviewer 4: abandoned and eligible for reprocessing ?

`cds.requires.scheduling` (multitenancy coordination):

| Option | Description |
|--------|-------------|
| `markerInterval` | Grid interval for markers; CAP picks a default that spreads tenant load across the interval |
| `flushInterval` | Cadence at which the central runner checks for tenants with pending work |

:::

::: details Configuration options for Java

> | Option | Default | Description |

Reviewer 4: Java example shows storeLastError: true, but storeLastError is not listed in the Java configuration options

|--------|---------|-------------|
| `maxAttempts` | `10` | Maximum retries before the entry is considered dead |
| `ordered` | `true` | Process entries in submission order |

:::

#### Disabling the Queue

You can disable event queues globally:

```json
{
  "cds": {
    "requires": {
      "queue": false
    }
  }
}
```

Or disable queueing for a specific service:

```json
{
  "cds": {
    "requires": {
      "messaging": {
        "outboxed": false
      }
    }
  }
}
```

## How It Works

### The Data Model

The persistent queue stores its messages in this entity, which CAP adds to your model on build and deploys with your other entities:

```cds
namespace cds.outbox;

entity Messages {
  key ID                   : UUID;           // Unique message identifier
      timestamp            : Timestamp;      // When the message was queued
      target               : String;         // Target service/queue name
      msg                  : LargeString;    // Serialized event payload
      attempts             : Integer default 0;  // Number of processing attempts

>       partition            : Integer default 0;

Reviewer 4: add comment like "// reserved"

      lastError            : LargeString;    // Error from last failed attempt
      lastAttemptTimestamp : Timestamp;      // When last attempt occurred
      status               : String(23);     // Current processing status
      task                 : String(255);    // Task name for named/singleton tasks

>       appid                : String(255);    // Application ID for shared HDI containers

Reviewer 4: The comment is quite terse. Consider expanding it to clarify the purpose or add a sentence after the data model explaining like "appid enables multiple applications to share the same outbox table without task name collisions"

CAP spreads marker timestamps across tenants so that processing doesn't synchronize into bursts — you don't need to configure that.

## Working with Event Queues

This section covers what you need to know to operate an event queue in production: how errors are retried, how to manage stuck messages, and how authorization carries over from the original request.

### Inspecting the Queue

To see what's currently queued for a tenant, query `cds.outbox.Messages` directly. The columns most useful for triage are `status`, `attempts`, `target`, `lastError`, and `lastAttemptTimestamp`:

```sql
SELECT ID, target, status, attempts, lastAttemptTimestamp, lastError
  FROM cds_outbox_Messages
  ORDER BY timestamp DESC;
```

For a managed view with bound *revive* and *delete* actions, expose a CDS service over the same entity — see [Dead Letter Queue](#dead-letter-queue) below. The same projection can be widened (drop the `attempts >= maxAttempts` filter) to inspect *all* pending messages, not just dead letters.

### Error Handling

When processing fails, the system retries the message with exponentially increasing delays.

> After a configurable maximum number of attempts, the message is moved to the dead letter queue.

Reviewer 4: not moved, just considered as dead message

Some errors are identified as _unrecoverable_ — for example, when a topic is forbidden by the broker.
These messages are immediately moved to the dead letter queue without further retries.

::: details When is a message picked up next?
A pending message is *processable* when all three conditions hold:

> 1. Its scheduled timestamp plus the retry backoff (`attempts × <exponential factor>`) is in the past.

Reviewer 4: The notation attempts × <exponential factor> is  potentially misleading. The number of attemps is already in <exponential factor> - no need to multiply. If multiplying, then it looks like linear function. 

exponential_factor(attempts)

or simply

exponential backoff based on attempts

    actions {
      action revive();
      action delete();
    };

}
```
:::

::: warning Restrict access
The dead letter queue contains sensitive data.
Ensure the service is accessible only to internal users.
:::

##### 2. Filter for Dead Entries

As `maxAttempts` is configurable, its value cannot be added as a static filter to the projection, but must be applied programmatically.

::: code-group
```js [Node.js — srv/outbox-dead-letter-queue-service.js]
const cds = require('@sap/cds')

module.exports = class OutboxDeadLetterQueueService extends cds.ApplicationService {
  async init() {
    this.before('READ', 'DeadOutboxMessages', function (req) {
      const { maxAttempts } = cds.env.requires.queue
      req.query.where('attempts >= ', maxAttempts)
    })
    await super.init()
  }
}
```
```java [Java — DeadOutboxMessagesHandler.java]
@Component
@ServiceName(OutboxDeadLetterQueueService_.CDS_NAME)
public class DeadOutboxMessagesHandler implements EventHandler {

  private final PersistenceService db;

  public DeadOutboxMessagesHandler(
      @Qualifier(PersistenceService.DEFAULT_NAME) PersistenceService db) {
    this.db = db;
  }

>   @Before(entity = DeadOutboxMessages_.CDS_NAME)

Reviewer 4: Should be:
@Before(event = CqnService.EVENT_READ, entity = DeadOutboxMessages_.CDS_NAME)

The handler takes CdsReadEventContext and modifies a CqnSelect query, so it's clearly intended for READ operations. Without event, the annotation matches all events, which is misleading.

  public void addDeadEntryFilter(CdsReadEventContext context) {
    Optional<Predicate> outboxFilters = createOutboxFilters(context.getCdsRuntime());
    outboxFilters.ifPresent(filter -> {
      CqnSelect modified = copy(context.getCqn(), new Modifier() {
        @Override
        public CqnPredicate where(Predicate where) {
          return filter.and(where);
        }
      });
      context.setCqn(modified);
    });
  }
}
```
:::

##### 3. Implement Bound Actions

Entries in the dead letter queue can be _revived_ by resetting the retry counter to zero, or _deleted_ permanently.

::: code-group
```js [Node.js — srv/outbox-dead-letter-queue-service.js]
this.on('revive', 'DeadOutboxMessages', async function (req) {
  await UPDATE(req.subject).set({ attempts: 0 })
})

this.on('delete', 'DeadOutboxMessages', async function (req) {
  await DELETE.from(req.subject)
})
```
```java [Java]

> @On

Reviewer 4: just @On, not @On(event = "...") ?

public void reviveOutboxMessage(DeadOutboxMessagesReviveContext context) {
  CqnAnalyzer analyzer = CqnAnalyzer.create(context.getModel());
  Map<String, Object> key = analyzer.analyze(context.getCqn()).rootKeys();
  Messages msg = Messages.create((String) key.get(Messages.ID));
  msg.setAttempts(0);
  db.run(Update.entity(Messages_.class).entry(key).data(msg));
  context.setCompleted();
}

> @On

Reviewer 4: just @On, not @On(event = "...") ?
